src/dmd/dsymbolsem.d

```diff
@@ -2108,7 +2108,117 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
-        return attribSemantic(uad);
+        attribSemantic(uad);
+        if (uad.atts)
+        {
+            // detect the special attribute @__gnu_abi_tag(...)
+            Strings gnuAbiTags;
+            bool checkDecl;
+            foreach (exp; *uad.atts)
+            {
+                bool noArgs;
+                if (auto ie = exp.isIdentifierExp())
+                {
+                    if (ie.ident == Id.gnuAbiTag)
+                    {
+                        noArgs = true;
+                        checkDecl = true;
+                        ie.type = Type.tvoid;
+                    }
+                }
+                else if (auto ce = exp.isCallExp())
+                {
+                    if (auto ie = ce.e1.isIdentifierExp())
+                    {
+                        if (ie.ident == Id.gnuAbiTag)
+                        {
+                            arrayExpressionSemantic(ce.arguments, sc);
+                            expandTuples(ce.arguments);
+                            noArgs = !ce.arguments.dim;
+                            Largs: foreach (i, arg; *ce.arguments)
+                            {
+                                OutBuffer buf;
+                                buf.printf("`@%s()`", Id.gnuAbiTag.toChars());
+                                auto se = semanticString(sc, arg, buf.extractString());
+                                if (!se)
+                                {
+                                    uad.errors = true;
+                                    continue;
+                                }
+                                (*ce.arguments)[i] = se;
+                                if (se.sz != 1)
+                                {
+                                    se.error("`@%s()` characters can only be of type `char`", Id.gnuAbiTag.toChars());
+                                    uad.errors = true;
+                                    continue;
+                                }
+                                foreach (c; se.peekSlice())
+                                {
+                                    if (!c.isValidMangling())
+                                    {
+                                        se.error("`@%s()` char 0x%02x not allowed", Id.gnuAbiTag.toChars(), c);
+                                        uad.errors = true;
+                                        continue Largs;
+                                    }
+                                }
+                                gnuAbiTags.push(se.toStringz().ptr);
+                            }
+                            ce.type = Type.tvoid;
+                            checkDecl = true;
+                        }
+                    }
+                }
+                if (noArgs)
+                {
+                    error(exp.loc, "`@%s()` at least one argument expected of type `string`", Id.gnuAbiTag.toChars());
+                    uad.errors = true;
```

Comment:
All this extra checking will be handled automatically if implemented as a true UDA.